### ISSUE TITLE
Seeking in audio CMAF track that uses an edit list and 1 or more priming samples

### ISSUE THREAD

**Issue by RufaelDev** [MPEGGroup/CMAF#24]
It would be good to clarify the behaviour when seeking in a CMAF track with MP4 audio that uses an edit list and 1 or more priming samples. 

Questions that come to mind
1. does the renderer have to apply the priming/preroll again or not ? 

Clarifying this would greatly help users of CMAF Tracks with MP4 audio

**Comment by cconcolato** [comment]
I don't think this is a CMAF question, more an ISOBMFF question.

First, I see a difference between ISOBMFF samples that contain "priming audio samples" (only at the beginning of the stream) and the fact that you need some previous ISOBMFF samples when starting at an ISOBMFF sample to avoid fade-in. We should be using 2 different terms to avoid confusion. I assume "priming" for the first and "preroll" for the second.

ISOBMFF does not (and in my view should not mention priming). ISOBMFF samples that contain "priming audio samples" are treated as any other samples (including potentially edited out with edit lists). There is no reason to apply them (I wouldn't know what it means) anywhere else than where they are present. 

The terminology from ISOBMFF is unfortunately ambiguous: there is a `AudioRollRecoveryEntry` (with 4CC `roll`) which:
> documents the pre-roll distance required in audio streams in which
every sample can be independently decoded, but the decoder output is only assured to be correct after
pre-rolling by the indicated number of samples.

and there is an `AudioPreRollEntry` (with 4CC `prol`) which:
> should be used with audio streams in which not every sample can be independently decoded

that is not used in CMAF.

CMAF uses "priming" in one place (10.3.3.1), which I think is not ideal because it mixes it with the use of the `roll` sample group, which is related to "preroll". Maybe we could rephrase that sentence to avoid using term "priming".  Maybe we could add a sentence about "priming" to say that it should be avoided. I don't think we can mandate it not to be used as lots of content use it (unfortunately).

Applying preroll (if properly signaled, as allowed in CMAF) is a decision of the player. Neither ISOBMFF nor CMAF mandate player behavior and they should not. CMAF already mentions that you can use `roll` sample groups. I'm not sure what else is needed.

**Comment by wschidol** [comment]
AS @cconcolato mentions, applying the preroll is optional for the player to do. As such, I think that the edit list applies to the decoder output without preroll applied. In other words, if preroll results in N additional samples then the edit list must be adjusted to cut these N samples away in addition to what it already implies.
I reality, I am not sure how common it is for audio formats to contain preroll data, and how common it is for realtime players/decoder to actually do preroll. @RufaelDev do you have any data on this?

**Comment by RufaelDev** [comment]
Thanks very much for the answers. The reason I asked in CMAF because the main use case is for (fragmented) audio tracks as documented in CMAF and to get feedback from audio experts, as we understand the ISOBMFF aspects very well but not the envisioned/ intended usage in audio CMAF tracks. We see such edits in commonly used encoders for different mp4 audio formats like aac.  We have seen encoders producing edit lists at the beginning for removing some initial (silent) samples from the presentation time line. Applying the edit at the beginning of the presentation is fine, but we wanted to get some clarification on the seeking behaviour. A direct interpretation of ISOBMFF would imply **not** getting prior samples again to do a pre-roll while seeking as there is no dependency between samples. 
Clause CMAF G.5. gives some guidance on the seeking and behaviour which is already very helpfull, as it recommend this fade-in approach. That makes a lot of sense. 

But @wschidol this slightly contradicts what you said as, when I seek, instead a fade-in or silence is applied without extra samples. This was I guess our main question and also if this fade-in approach can be used at the beginning of the track (in many cases such as live there is not really an easily determined beginning and this may be the better default approach). While this solves the issue, we feel in practice indeed edit list may still be used to re-apply pre-roll without fade-in in real-time and live cases when seeking or tuning in, which is strictly speaking not a 100% correct interpretation of ISOBMFF. So perhaps documenting some of that may still be helpful.

**Comment by wschidol** [comment]
I am somewhat confused by the connection between edit lists and pre-roll. One can have edit lists w/o pre-roll and vice versa, so they don't *have* to come together.
Note, when I speak of pre-roll I describe that which is signaled in 14496-12 clause 10.1.1. What CMAF G.5 describes is **not** pre-roll as per clause 14496-12 10.1.1. 14496-19 G.5 speaks about latency caused by the encoder. This latency would be signaled and compensated through an **edit list**, not communicated as pre-roll. The "fade-in" that is described refers to a form of concealment; it is not something one should rely on or purposefully trigger -- it is a best-effort approach for an audio codec to output audio when it was not started up with an i-frame.
The implementation suggested in G.5 is not a recommendation and not necessarily the best way to go about this as it will potentially result in dropping the initial i-frame. For example, an alternative to G.5 would be to drop not the **encoded** AUs but the **decoded** AUs.
All that said, I am not sure where you see the contradiction?

**Comment by RufaelDev** [comment]
I want to thank you all again for considering our issues raised around MP4 audio,
while we understand CMAF may not solve all issues related to (legacy) MP4 audio, it may 
be very helpful to have some informative guidance on playback and authoring 
of MP4 audio CMAF tracks. This may help other organizations that are 
developing specific streaming solutions using CMAF and MP4 audio. 

To lead a bit into the discussion, 
I provide some of the concepts and goals as also initiated
on the respective github issue: 

Audio codec tools:
priming: 
Some extra samples inserted at the beginning of the stream, 
may involve decoding samples not intended for presentation,
explained in Annex G.5

preroll -> operation to re-initialize 
the audio decoder for correct sample reconstruction, 
may use prior samples or only current samples and a fade-in to mask
the perceptual transition phase. Also mentioned in Annex G.5 

ISO-BMFF tools: 
edit-list 
prol box  

Key streaming use cases to discuss: 

1. VoD (he-)aac/ec-3 with and without an edit list 

a) starting a VoD clip from start 
b) starting a VoD clip from a certain presentation time 
c) Seeking in a VoD clip (the clip was already playing)
d) transitioning to an ad break 

2. Live (he-)aac/ec-3 with and without an edit list
a) joining a live stream from the start 
b) joining a live stream while it is playing (i.e. we cannot really insert an edit every time)
c) seeking in a DVR window of a live stream 
d) transitioning to an ad break 

Some key goals: 

- avoid audio glitches at transitions 
- audio/video segment boundary alignment
- avoid sync issues, for lower sample rates and for example HE-AAC, 
  2 samples at 24 KhZ we are already talking about ~85 ms 
- as good audio quality/reconstruction as possible
- consistent seeking
- consistent authoring

**Comment by dwsinger** [comment]
to summarize:
* the edit list moves the times so that the audio is sync'd with other media
* the 'roll' (not prol) sample group tells you that IF you want sample-correct audio, you should pre-roll the decoder, and how much by (in file format frames, which is all you have at that point); it applies to all frames in the stream
* there are extra audio samples present in the stream before the first audio sample of the program material, to enable this pre-roll at the stream start
* the edit list documents how many extra audio samples there are, by explicitly cutting them out of the output
* the edit list also can document when the program material doesn't end on a frame boundary

so, pre-rolling is optional; if you don't, you'll get a 'soft entry' (cross-fade or fade-in). 

If priming samples are present in the audio signal, the edit list is needed to get sync (and hence effectively conditionally mandatorially present).

If the program audio signal does not end on a frame boundary, the edit list is recommended to document the last program audio sample.

If the edit list is present, you must obey it in the reader, or you'll be out of sync.

The edit list also documents how to round-trip from raw audio of N samples back into a raw audio of N samples after encoding, packaging, unpacking, and decoding.

**Comment by cconcolato** [comment]
@dwsinger I almost agree with all you said but the mandatory edit list. Yes, adding 'priming' samples and documenting them through an edit list is necessary to guarantee round-trip, but I'm told encoders can avoid adding these 'priming' samples. If you don't, you'll get a 'soft entry' also at the beginning but you'll avoid the need for an edit list, still guaranteeing sync (at least in compliant players).

**Comment by dwsinger** [comment]
I don't mean to imply its presence is mandatory; I meant that if it's there, it's mandatory to handle it (reader side). sorry. I will edit the comment to say that.

I totally agree, one really clean solution (that I would advocate) is not to add priming samples to audio signals, and instead have the first sample out of the first frame be the first program audio sample, and you'll get a fade-in. Most audio signals start with silence anyway, and a fade-in to silence is really hard to notice.

**Comment by RufaelDev** [comment]
Thanks so much for this summary @dwsinger really helpful for everyone working with this technology, 

indeed I also captured this last option from the discussions we had with audio experts, thanks again.

**Comment by haudiobe** [comment]
I also had some offline chat with Cyril on this one. I created three slides on my end. I tried to document the exact timing aspects. Very important is the test of MSE implementations. I wrote down a few tests, but one or two on time alignment still need to be fully defined.
https://mpeg.expert/live/nextcloud/index.php/s/zszYys25k5TZzPM

**Comment by haudiobe** [comment]
I did one more update on this with expected MSE behaviour and different options for a client operation. I believe we need to do a few MSE tests. Transition is important, but most important is time-sync.

[Audio-and-DASH.pptx](https://github.com/MPEGGroup/CMAF/files/6399107/Audio-and-DASH.pptx)

Linked also on nextcloud

**Comment by wschidol** [comment]
Agree with most of what is said except that "[audio] encoders can avoid adding these 'priming' samples.". All audio encoders have lookahead, some more, some less, and in a causal system that lookahead will show up as "additional" samples at the encoder output (which I would call delay or latency -- I think @dwsinger might prefer latency).
What can be done in a system with audio and video encoder is to discard the additional compressed samples encoder-side and start the audio segment late. To get to same-duration segments, the audio segment also needs to finish "late" which means that the compressed video segment will have to be delayed by the same time before transmission. In DASH, that "delay" would really only manifest itself in when the first segment of a period is available.
This might be a fine point -- but just to call out that an encoder/packager system does not come for free, in terms of delay and engineering effort. So we should not assume that edit lists are the exception.

I also agree with you @haudiobe that more tests will be useful although I admit that I did not understand your slide in detail.

To sum up, as long as we end up with...

- if an edit list is present, you must obey it in the reader (as already pointed out by @dwsinger )
- even if it may not be preferred, encoders can write edit lists and rely on defined behavior in DASH players
- writing roll elements is optional
- applying roll elements (i.e., pre-roll) is optional for the decoder

... we (Dolby) are fine. I believe none of these points contradict what has been agreed above.

**Comment by cconcolato** [comment]
We are closing this issue but if people think the CMAF specification needs to be updated, we welcome contribution text based on this discussion.
